projects/dataset_recorder/auto_dataset_recorder.cc

```diff
@@ -0,0 +1,174 @@
+      : m_Catcher{ catcher }
+      , m_HeadingOffset{ calculateHeadingOffset() }
+      , m_Recorder{ m_GPS, useSystemClock }
+      , m_Controller{ oldDatabase, LookAheadDistance,
+                      m_Robot.getDistanceBetweenAxes(), StoppingDist }
+    {}
+
+    void run()
+    {
+        /*
+         * Drive the robot along the old route, stopping only when:
+         * 1. The robot has reached its destination
+         * 2. The PurePursuitController is "stuck"
+         * 3. Ctrl+C is pressed
+         */
+        Pose2<millimeter_t, radian_t> pose;
+        LOGI << "RECORDING STARTED";
+        do {
+            checkForErrors();
+
+            // Record sensor values to database
+            m_Recorder.step(m_Robot, m_IMU);
+
+            // Get the robot's current pose from GPS + IMU readings
+            const auto &gpsData = m_Recorder.getGPSData();
+            const auto xy = MapCoordinate::latLonToUTM(gpsData.coordinate).toVector();
+            pose.x() = xy.x();
+            pose.y() = xy.y();
+            pose.yaw() = normaliseAngle180(m_IMU.getEulerAngles()[0] + m_HeadingOffset);
+        } while (m_Controller.step(pose, m_Robot, MaxSpeed, MaxTurn));
+    }
+
+private:
+    Robots::Ackermann::RCCarBot m_Robot;
+    GPS::GPSReader m_GPS;
+    BN055 m_IMU;
+    const BackgroundExceptionCatcher &m_Catcher; // Catch Ctrl+C etc.
+    const radian_t m_HeadingOffset;              // Offset between GPS and IMU headings
+    RobotRecorder m_Recorder;                    // Helper to record sensor data to ImageDatabase
+    AutoController m_Controller;                 // For automatically driving the robot
+
+    void checkForErrors()
+    {
+        // Background exceptions/signals
+        m_Catcher.check();
+
+        // Allow user to abort with remote control
+        const auto move = m_Robot.readRemoteControl();
+        if (move.first >= SpeedAbortThresh || move.second >= TurnAbortThresh) {
+            throw std::runtime_error{ "Control overriden with remote control; aborting" };
+        }
+    }
+
+    radian_t calculateHeadingOffset()
+    {
+        constexpr float InitialDriveSpeed = 0.5f;
+        constexpr meter_t InitialDriveDistance = 1_m;
+        constexpr auto DriveTimeout = 20s;
+        constexpr auto ReadDelay = 50ms;
+
+        LOGI << "Calculating robot's position with GPS...";
+
+        // Get initial position (try to get reading 5 times)
+        const MapCoordinate::UTMCoordinate utmStart = [&]() {
+            LOGI << "Waiting for initial GPS fix";
+            for (int i = 0; i < 5; i++) {
+                checkForErrors();
+
+                const auto data = m_GPS.read();
+                if (data && data->gpsQuality != GPS::GPSQuality::INVALID) {
+                    return MapCoordinate::latLonToUTM(data->coordinate);
+                }
+
+                std::this_thread::sleep_for(ReadDelay);
+            }
+
+            throw std::runtime_error{ "Could not get initial GPS fix" };
+        }();
```

Comment:
why is this a lambda function? Just so you can make ``utmStart`` const?

Reply:
Partly. But it does also make the logic a bit clearer because we can do a ``return`` in that loop body. An alternative would be to turn this into a separate member function.